src/sss_client/pam_sss.c

```diff
@@ -1939,43 +1939,56 @@ static int prompt_multi_cert(pam_handle_t *pamh, struct pam_items *pi)
     return ret;
 }
 
+#define SC_INSERT_PROMPT "Please (re)insert (different) Smartcard"
+
 static int prompt_sc_pin(pam_handle_t *pamh, struct pam_items *pi)
 {
     int ret;
     char *answer = NULL;
-    char *prompt;
-    size_t size;
+    char *prompt = NULL;
+    size_t size = 0;
     size_t needed_size;
     const struct pam_conv *conv;
     const struct pam_message *mesg[2] = { NULL, NULL };
     struct pam_message m[2] = { { 0 }, { 0 } };
     struct pam_response *resp = NULL;
     struct cert_auth_info *cai = pi->selected_cert;
-    struct cert_auth_info empty_cai = { NULL, NULL, discard_const("Smartcard"),
-                                        NULL, NULL, NULL, NULL, NULL, NULL,
-                                        NULL, NULL };
 
-    if (cai == NULL && SERVICE_IS_GDM_SMARTCARD(pi)) {
-        cai = &empty_cai;
+    if (cai == NULL && (SERVICE_IS_GDM_SMARTCARD(pi)
+                            || (pi->flags & PAM_CLI_FLAGS_REQUIRE_CERT_AUTH))) {
+        size = sizeof(SC_INSERT_PROMPT);
```

Comment:
Imo, logic could be simplified using `asprintf()` in place (right here) instead of `size` + `malloc` + `snprintf`...

Reply:
changed